bndtools.builder/src/org/bndtools/builder/BndtoolsBuilder.java

```diff
@@ -231,8 +233,8 @@
 						force = true;
 					}
 
-					if (!force && delta.hasNoTarget(model)) {
-						buildLog.basic("project has no target files");
+					if (!force && delta.hasOutOfDateTarget(model, lastModifiedClasspathContainer(myProject))) {
+						buildLog.basic("project has no or out-of-date target files");
```

Comment:
Ok, I've had a bit of think about this and I have one minor nit. I prefer log messages to be precise. Rather than test for multiple conditions and issue a generic leg message when any of them fail, I think it's better to check each one individually and log an error message specific to the failure.

Log messages like these are only useful rarely. However, in those rare situations where they do come in useful, usually the more precise they are the more useful they become. Especially if they are the result of an intermittent failure that you can't readily reproduce. If a build is occasionally doing something unexpected and this message appears in the log, the person looking at it may wish that they knew whether a rebuild was triggered because the file was out-of-date, or else if it triggered because it didn't exist.

So my preference would be to check for "no target" and "out-of-date target" as two separate checks, and log each of them separately as appropriate.

Reply:
> I think it's better to check each one individually and log an error message specific to the failure.

But this is just duplicating the loop execution. It is the same loop to look for whether the target files are missing or out-of-date. While having a precise log message is nice, it is not worth duplicating code/work to get there. It is just a log message.

In a way, a missing file is a special case of being out-of-date (which is what the code in the PR reflects). So I would  be more inclined to change the log message to just say out-of-date and not mention missing.